### PR TITLE
repart: Switch to new mkfs.btrfs subvolume API

### PR DESCRIPTION
**Pull request by daandemeyer**
In https://github.com/kdave/btrfs-progs/pull/877 the API is changing
to "--subvol <path>:ro,default" so let's adapt our usage to match.

This also adds support for read-only subvolumes.

Fixes https://github.com/systemd/systemd/issues/34134

### CODE REVIEW

File: src/partition/repart.c [hunk 1]
```diff
@@ -259,6 +260,81 @@ static PartitionEncryptedVolume* partition_encrypted_volume_free(PartitionEncryp
...
 
+typedef enum SubvolumeFlags {
+        SUBVOLUME_DEFAULT          = 1 << 0,
+        SUBVOLUME_RO               = 1 << 1,
+        _SUBVOLUME_FLAGS_MASK      = SUBVOLUME_DEFAULT|SUBVOLUME_RO,
+        _SUBVOLUME_FLAGS_INVALID   = -EINVAL,
+        _SUBVOLUME_FLAGS_ERRNO_MAX = -ERRNO_MAX, /* Ensure the whole errno range fits into this enum */
+} SubvolumeFlags;
+
+static SubvolumeFlags subvolume_flags_from_string_one(const char *s) {
```
Comment (yuwata) [inline src/partition/repart.c:270]: `assert(s);`

File: src/partition/repart.c [hunk 2]
```diff
@@ -1677,6 +1752,87 @@ static int config_parse_make_dirs(
...
+                                continue;
+                        }
+                        if (flags < 0) {
+                                log_syntax(unit, LOG_WARNING, filename, line, r, "Failed to parse subvolume flags, ignoring: %s", f);
+                                continue;
+                        }
+
+                        s->flags = flags;
+                }
+
```
Comment (yuwata) [inline src/partition/repart.c]: Unnecessary empty line.

File: src/partition/repart.c [hunk 3]
```diff
@@ -2087,9 +2245,26 @@ static int partition_read_definition(Partition *p, const char *path, const char
...
                                   verity_mode_to_string(p->verity));
 
-        if (p->default_subvolume && !path_strv_contains(p->subvolumes, p->default_subvolume))
-                return log_syntax(NULL, LOG_ERR, path, 1, SYNTHETIC_ERRNO(EINVAL),
-                                  "DefaultSubvolume= must be one of the paths in Subvolumes=.");
+        if (default_subvolume) {
+                subvolume = ordered_hashmap_get(p->subvolumes, default_subvolume);
+                if (!p)
+                        return log_syntax(NULL, LOG_ERR, path, 1, SYNTHETIC_ERRNO(EINVAL),
+                                        "DefaultSubvolume= must be one of the paths in Subvolumes=.");
```
Comment (yuwata) [inline src/partition/repart.c]: Indentation.

**Review by yuwata — APPROVED** [review]
LGTM. Several minor comments.

**Comment by DaanDeMeyer** [comment]
Let's not merge this until the corresponding PR in btrfs-progs is merged

File: man/repart.d.xml [hunk 4]
```diff
@@ -541,33 +541,43 @@
...
+              <row>
+                <entry>Flag</entry>
+                <entry>Purpose</entry>
+              </row>
+            </thead>
+            <tbody>
+              <row id='D'>
+                <entry><literal>default</literal></entry>
+                <entry>Make this subvolume the default subvolume. Can only be specified for a single subvolume.</entry>
+              </row>
```
Comment (poettering) [inline man/repart.d.xml]: hmm, this is weird. This is not really a flag, it's a singleton after all. kinda weird exposing it like this

Comment (daandemeyer) [reply man/repart.d.xml]: OK we can stick with the previous solution for the default subvolume then.

File: src/shared/image-policy.c [hunk 5]
```diff
@@ -437,7 +437,7 @@ int partition_policy_flags_to_string(PartitionPolicyFlags flags, bool simplify,
         if (m == 0)
                 buf = strdup("-");
         else {
-                assert(m+1 < ELEMENTSOF(l));
+                assert(m < ELEMENTSOF(l));
```
Comment (poettering) [inline src/shared/image-policy.c:440]: commit msg is misleading. not a null byte, but a null pointer.

File: src/partition/repart.c [hunk 6]
```diff
@@ -2089,7 +2242,7 @@ static int partition_read_definition(Partition *p, const char *path, const char
                                   "SizeMinBytes=/SizeMaxBytes= cannot be used with Verity=%s.",
                                   verity_mode_to_string(p->verity));
 
-        if (p->default_subvolume && !path_strv_contains(p->subvolumes, p->default_subvolume))
+        if (p->default_subvolume && !ordered_hashmap_contains(p->subvolumes, p->default_subvolume))
                 return log_syntax(NULL, LOG_ERR, path, 1, SYNTHETIC_ERRNO(EINVAL),
                                   "DefaultSubvolume= must be one of the paths in Subvolumes=.");
```
Comment (poettering) [inline src/partition/repart.c:2247]: i wonder if it wouldn't be nicer to simply add the default subvol to the existing subvolumes if missing, with default settings. but doesn't matter

File: src/partition/repart.c [hunk 7]
```diff
@@ -5025,20 +5186,47 @@ static int do_copy_files(Context *context, Partition *p, const char *root) {
...
+        _cleanup_strv_free_ char **subvolumes = NULL;
+        Subvolume *subvolume;
         int r;
 
         assert(p);
         assert(root);
 
+        ORDERED_HASHMAP_FOREACH(subvolume, p->subvolumes)
+                if (strv_extend(&subvolumes, subvolume->path) < 0)
+                        return log_oom();
```
Comment (poettering) [inline src/partition/repart.c]: add a helper for this?

Comment (poettering) [reply src/partition/repart.c]: you have the same loop twice after all

**Comment by DaanDeMeyer** [comment]
The corresponding PR in mkfs.btrfs was merged, so merging this as well

**Pull request merged**